### ISSUE TITLE
Azure Storage should have the option to use a secret name for the connection string

### ISSUE THREAD

**Issue by zangell44** [PrefectHQ/prefect#4447]
## Current behavior
Right now, the connection string to Azure can be passed as a string at initialization or read `AZURE_STORAGE_CONNECTION_STRING` from the environment.

The connection string property is not serialized with the storage object. The only way to get this to work is to have `AZURE_STORAGE_CONNECTION_STRING` available when the flow is retrieved from storage. For most agent types, this means passing it as an environment variables (1) via cli command on agent start or (2) via run config

## Proposed behavior

- Azure storage should optionally take a new parameter `connection_string_secret_name`, which will be serialized with the storage model
- If the parameter is not None, the secret will be used to populate `self.connection_string`
- If the parameter is None, we should fall back to the old environment variable loading behavior (this should not be a breaking change)
- We should remove explicit passing of `connection_string`, which is "breaking"-ish, but this doesn't really work imo

## Example

This will simplify and make explicit the use of connection strings for Azure storage.

It'll look something like this
```python
# in prefect.storage.azure.py
class Azure(Storage):
    ....
    @property
    def connection_string(self) -> str:
        if self.connection_string_secret_name is not None:
            return Secret(self.connection_string_secret_name).get()

       connection_string = os.getenv("AZURE_STORAGE_CONNECTION_STRING")
       if connection_string is None:
            raise Exception("Azure connection string not provided ... < more useful info>")
       return connection_string
       
    @property
    def _azure_block_blob_service(self):  # type: ignore
        import azure.storage.blob
        return azure.storage.blob.BlobServiceClient.from_connection_string(
            conn_str=self.connection_string
        )

# example usage
from prefect import Flow
from prefect.storage import Azure

storage = Azure(
    container="my-container", 
    blob_name="my-blob", 
    connection_string_secret_name="SUPAH_SECRET_AZURE_CONNECTION"
)

# secret name will be present when the flow is retrieved
assert storage.serialize()["connection_string_secret_name"] == "SUPAH_SECRET_AZURE_CONNECTION"

flow = Flow('my-azure-flow')
flow.storage = storage
```

**Comment by ciaranevans** [comment]
@zangell44 I just want to confirm your statement of:

```
The only way to get this to work is to have AZURE_STORAGE_CONNECTION_STRING present in the Agent environment when a flow is retrieved.
```

I've deployed a Prefect Agent with this value available, but it still falls over on Flow running, I believe it's the runner of the flow that needs this Environment Variable, _not_ the Agent. 

For example, my Agent has the value of `AZURE_STORAGE_CONNECTION_STRING` populated, and my flow run is setup like so:

```python
run_config=KubernetesRun(
    image="prefecthq/prefect:0.14.17-python3.8",
    labels=json.loads(os.environ["PREFECT__CLOUD__AGENT__LABELS"]),
),
```

And this gives me:

```
Failed to load and execute Flow's environment: AttributeError("'NoneType' object has no attribute 'rstrip'")
```

However, if I change my run config to have the Environment Variable, and remove the Environment Variable from my Agent:

```python
run_config=KubernetesRun(
    image="prefecthq/prefect:0.14.17-python3.8",
    env={"AZURE_STORAGE_CONNECTION_STRING": os.environ["FLOW_STORAGE_CONNECTION_STRING"]},
    labels=json.loads(os.environ["PREFECT__CLOUD__AGENT__LABELS"]),
),
```

I get a successful flow run:

![image](https://user-images.githubusercontent.com/9111975/116395279-d4a37e80-a81b-11eb-858c-0c596d71b07f.png)

**Comment by zangell44** [comment]
@ciaranevans good catch! You're correct, it's a bit trickier than I originally put it. I'll update the description.

A better rephrasing is "The only way to get this to work is to have AZURE_STORAGE_CONNECTION_STRING present in the environment where the flow is retrieved from storage".

The mechanics of flow retrieval vary slightly across agents. 

When creating a job spec, the Kubernetes Agent includes environment variables (1) that were passed to the agent via the cli command and (2) that were passed to the Agent via RunConfig for the flow run.

Similar patterns are followed by other agents. `LocalAgent` is actually the only one which copies all existing environment variables, so my original statement is only correct for this agent.


TL;DR - if you're running into this issue right now, you can make it work by passing `AZURE_STORAGE_CONNECTION_STRING` via (1) agent start cli command or (2) run config environment variables

**Comment by ciaranevans** [comment]
Cool, no worries @zangell44! Yeah for now the workaround we have is to put it in the run config 👍
